controllers/machineset_controller.go

```diff
@@ -644,12 +670,19 @@ func (r *MachineSetReconciler) updateStatus(ctx context.Context, cluster *cluste
 		newStatus.DeepCopyInto(&ms.Status)
 
 		log.V(4).Info(fmt.Sprintf("Updating status for %v: %s/%s, ", ms.Kind, ms.Namespace, ms.Name) +
-			fmt.Sprintf("replicas %d->%d (need %d), ", ms.Status.Replicas, newStatus.Replicas, *ms.Spec.Replicas) +
+			fmt.Sprintf("replicas %d->%d (need %d), ", ms.Status.Replicas, newStatus.Replicas, desiredReplicas) +
 			fmt.Sprintf("fullyLabeledReplicas %d->%d, ", ms.Status.FullyLabeledReplicas, newStatus.FullyLabeledReplicas) +
 			fmt.Sprintf("readyReplicas %d->%d, ", ms.Status.ReadyReplicas, newStatus.ReadyReplicas) +
 			fmt.Sprintf("availableReplicas %d->%d, ", ms.Status.AvailableReplicas, newStatus.AvailableReplicas) +
 			fmt.Sprintf("sequence No: %v->%v", ms.Status.ObservedGeneration, newStatus.ObservedGeneration))
 	}
+	if newStatus.Replicas >= desiredReplicas {
+		// This means that there was no error in generating the desired number of machine objects
```

Comment:
`updateStatus` is called regardless `syncReplicas` returns err or not.
`newStatus.Replicas` accounts for Machine CRs regardless the underlying infra was created or not.

This means than even though `MachinesCreatedCondition` was markFalse by `syncReplicas` it will always be wrongly set to true here. Am I missing anything?

Reply:
@enxebre Thank you for the review.

`syncReplicas` will return without creating the CR if there are issues during [bootstrap ](https://github.com/kubernetes-sigs/cluster-api/blob/764d1bd793a031e1c5f702a9c80dc248bf34832d/controllers/machineset_controller.go#L350 )and [infra](https://github.com/kubernetes-sigs/cluster-api/blob/764d1bd793a031e1c5f702a9c80dc248bf34832d/controllers/machineset_controller.go#L364) clone.

Similarly, the machine [create](https://github.com/kubernetes-sigs/cluster-api/blob/764d1bd793a031e1c5f702a9c80dc248bf34832d/controllers/machineset_controller.go#L368) can also fail but that should fine too as the CR will not have been created.

I also compared this to what KCP does and both look very similar.  See [cloneConfigsAndGenerateMachine](https://github.com/kubernetes-sigs/cluster-api/blob/7a75113b5d24528a36e1646354cd54f294b9320f/controlplane/kubeadm/controllers/helpers.go#L155)

Reply:
Q: what happens if we scale up a existing machine deployment/machine set? if I'm not wrong, with the current implementation the condition will remain true, even if additional machines are currently being created...

Reply:
In KCP, scaling is captured by the `ResizedCondition` and in that case, the `MachinesCreatedCondition` is [not set to false either](https://github.com/kubernetes-sigs/cluster-api/blob/f6fd5ed7dc0fa75788f76f17c64ae82976fcc70b/controlplane/kubeadm/controllers/status.go#L77-L79).

I suppose to be consistent, we should either keep this the code as is or change both KCP and MD to set `MachinesCreatedCondition` to false when `replicas < desiredReplicas`? Which would you prefer?

Reply:
Let's implement MS conditions similar to KCP, as in your original idea but adding resized condition too